
Issue: Lock-up during boot-up with fsa9480 DPLL constraints

Fix: 
	- Avoiding dpll_cascade enter/exit before abe initialization

	- Enter dpll cascading when all conditions are met:
          - system is in early suspend (screen is off)
          - single stream is active and is LP (ping-pong)
          - OPP is 50 or less (DL1 path only)
   
Index: kernel/arch/arm/mach-omap2/dpll-44xx.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/dpll-44xx.c	2011-07-13 21:36:24.891996408 +0530
+++ kernel/arch/arm/mach-omap2/dpll-44xx.c	2011-07-13 21:38:24.331998785 +0530
@@ -543,7 +543,7 @@
 
 	list_add(&blocker->node, &dpll_cascading_blocker_list);
 
-	if (list_was_empty && in_dpll_cascading) {
+	if (list_was_empty && in_dpll_cascading && abe_can_enter_dpll_cascading) {
 		printk("%s:%d Before DPLL exit: dev_name = %s\n",
 			__func__,__LINE__,dev_name(dev));
 		omap4_dpll_low_power_cascade_exit();
@@ -587,7 +587,7 @@
 	list_del(&blocker->node);
 
 	if (list_empty(&dpll_cascading_blocker_list)
-			&& !in_dpll_cascading) {
+			&& !in_dpll_cascading && abe_can_enter_dpll_cascading) {
 		in_dpll_cascading = true;
 		/* DPLL cascade stability check: Add printk for testing purpose. */
 		printk("%s:%d Before DPLL enter: dev_name = %s\n",
Index: kernel/arch/arm/mach-omap2/include/mach/omap4-common.h
===================================================================
--- kernel.orig/arch/arm/mach-omap2/include/mach/omap4-common.h	2011-07-13 21:36:49.951997981 +0530
+++ kernel/arch/arm/mach-omap2/include/mach/omap4-common.h	2011-07-13 21:38:24.331998785 +0530
@@ -100,6 +100,7 @@
 extern void *so_ram_address;
 
 extern bool in_dpll_cascading;
+extern bool abe_can_enter_dpll_cascading;
 extern rwlock_t dpll_cascading_lock;
 
 extern void __init gic_init_irq(void);
Index: kernel/sound/soc/omap/omap-abe-dsp.c
===================================================================
--- kernel.orig/sound/soc/omap/omap-abe-dsp.c	2011-07-13 21:37:38.012445260 +0530
+++ kernel/sound/soc/omap/omap-abe-dsp.c	2011-07-13 21:38:24.331998785 +0530
@@ -130,6 +130,8 @@
 #define ABE_DBG_FLAG2_SIZE	0
 #define ABE_DBG_FLAG3_SIZE	0
 
+bool abe_can_enter_dpll_cascading = false;
+
 /* TODO: fine tune for ping pong - buffer is 2 periods of 12k each*/
 static const struct snd_pcm_hardware omap_abe_hardware = {
 	.info			= SNDRV_PCM_INFO_MMAP |
@@ -2742,6 +2744,7 @@
 	abe->abe_pdata = pdata;
 	abe->pdev = pdev;
 	dpll_cascading_blocker_hold(&abe->pdev->dev);
+	abe_can_enter_dpll_cascading = true;
 
 	mutex_init(&abe->mutex);
 	mutex_init(&abe->opp_mutex);
